File: .claude/hooks/pre-compact-backup.mjs
#!/usr/bin/env node
// @hookstack pre-compact-transcript-backup
// Sauvegarde le résumé de compaction dans un fichier temporaire (PreCompact)
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join } from 'path';
import { fileURLToPath } from 'url';

export function run(
  input,
  {
    writeFile = writeFileSync,
    mkdir = mkdirSync,
    backupDir = '/tmp/claude-compact-backups',
    now = () => new Date().toISOString(),
  } = {},
) {
  const summary = input.summary ?? '';
  const sessionId = input.session_id ?? `session-${Date.now()}`;
  if (!summary) return null;

  mkdir(backupDir, { recursive: true });
  const file = join(backupDir, `${sessionId}.json`);
  writeFile(file, JSON.stringify({ session_id: sessionId, saved_at: now(), summary }, null, 2));

  return { file, message: `[pre-compact-backup] Contexte sauvegardé → ${file}\n` };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result?.message) process.stderr.write(result.message);
}

File: .claude/hooks/pre-compact-backup.py
#!/usr/bin/env python3
# @hookstack pre-compact-transcript-backup
"""Backs up the compaction summary to a temp file (PreCompact)."""
import json
import os
import sys
import time
from datetime import datetime, timezone


def _now():
    return datetime.now(timezone.utc).isoformat()


def _write_file(path, content):
    with open(path, "w", encoding="utf8") as f:
        f.write(content)


def _mkdir(path, *, recursive=True):
    os.makedirs(path, exist_ok=True)


def run(
    input_data,
    *,
    write_file=None,
    mkdir=None,
    backup_dir="/tmp/claude-compact-backups",
    now=None,
):
    if write_file is None:
        write_file = _write_file
    if mkdir is None:
        mkdir = _mkdir
    if now is None:
        now = _now

    summary = input_data.get("summary") or ""
    session_id = input_data.get("session_id") or f"session-{int(time.time() * 1000)}"
    if not summary:
        return None

    mkdir(backup_dir, recursive=True)
    file = os.path.join(backup_dir, f"{session_id}.json")
    payload = json.dumps(
        {"session_id": session_id, "saved_at": now(), "summary": summary},
        indent=2,
    )
    write_file(file, payload)

    return {"file": file, "message": f"[pre-compact-backup] Contexte sauvegardé → {file}\n"}


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result and result.get("message"):
        sys.stderr.write(result["message"])
